File: airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeSqlNameTransformerTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.snowflake;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;
import org.junit.jupiter.api.Test;

class SnowflakeSqlNameTransformerTest {

  private static final SnowflakeSQLNameTransformer INSTANCE = new SnowflakeSQLNameTransformer();
  private static final Map<String, String> RAW_TO_NORMALIZED_IDENTIFIERS = Map.of(
      "name-space", "name_space",
      "spécial_character", "special_character",
      "99namespace", "_99namespace");

  @Test
  public void testGetIdentifier() {
    RAW_TO_NORMALIZED_IDENTIFIERS.forEach((raw, normalized) -> {
      assertEquals(normalized, INSTANCE.convertStreamName(raw));
      assertEquals(normalized, INSTANCE.getIdentifier(raw));
      assertEquals(normalized, INSTANCE.getNamespace(raw));
    });
  }

}

File: airbyte-integrations/connectors/destination-snowflake/src/test/kotlin/io/airbyte/integrations/destination/snowflake/SnowflakeSqlNameTransformerTest.kt
/*
 * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.snowflake

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class SnowflakeSqlNameTransformerTest {

    private lateinit var transformer: SnowflakeSqlNameTransformer

    @BeforeEach
    fun setUp() {
        transformer = SnowflakeSqlNameTransformer()
    }

    @Test
    fun testToAlphanumericAndUnderscore() {
        assertEquals("users", transformer.transform("users"))
        assertEquals("users123", transformer.transform("users123"))
        assertEquals("UsErS", transformer.transform("UsErS"))
        assertEquals("users_USE_special_____", transformer.transform("users USE special !@#$"))
    }

    @Test
    fun testNameWithUnsupportedFirstCharacter() {
        val name = "-name-with-unsupported-first-character"
        assertEquals(
            "_${name.takeLast(name.length - 1).replace("-","_")}",
            transformer.transform(name)
        )
    }
}
